### ISSUE TITLE
map_entry warning does not consider possible intervening code

### ISSUE THREAD

**Issue by jmafc** [rust-lang/rust-clippy#5696]
Version: clippy 0.0.212 (d4092ac 2020-05-11)
I get the following warning:
```rust
172 | /         if !seen.contains_key(&url) {
173 | |             v.push(Href {url: url.clone(), title: row.get("title")});
174 | |             seen.insert(url, 1);
175 | |         }
    | |_________^ consider using `seen.entry(url)`
```
The warning doesn't take into account that the `contains_key()` / `insert()` sequence has some code that is necessary in between (it could also apply if the `push()` was placed after the `insert()`. For context, this is inside a loop iterating over rows and the Hashmap `seen` is used to prevent inserting duplicates into the vector `v`. AFAICT (I've tried several variations of using `Entry` methods), using the suggested `or_insert()` does *not* provide a solution.

**Comment by mikerite** [comment]
What about...
```rust
if let Entry::Vacant(v) = seen.entry(url) {
    v.push(Href {url: url.clone(), title: row.get("title")});
    entry.insert(1);
}
```
Now you're only looking up the key once so it should be faster. Aside from looking nicer.

**Comment by jmafc** [comment]
Yes, that works (except for having to use `url.clone()` in the first rather than the second one). I'm still wondering, though, whether the warning is always valid, irrespective of other code in the block following the `contains_key()`.

**Comment by mikerite** [comment]
There are cases. `entry()` mutably borrows the Hashmap so the suggestion is invalid if you're using the hashmap for something else in the block. See #4674.
